src/Compilers/Core/Portable/SourceGeneration/SourceGeneratorContext.cs

```diff
@@ -59,7 +60,30 @@ internal SourceGeneratorContext(Compilation compilation, ImmutableArray<Addition
 
         internal AdditionalSourcesCollection AdditionalSources { get; }
 
-        public void AddSource(string hintName, SourceText sourceText) => AdditionalSources.Add(hintName, sourceText);
+        /// <summary>
+        /// Adds a <see cref="SourceText"/> to the compilation
+        /// </summary>
+        /// <param name="hintName">An identifier that can be used to reference this source text, must be unique within this generator</param>
+        /// <param name="sourceText">The <see cref="SourceText"/> to add to the compilation</param>
+        public void AddSource(string hintName, SourceText sourceText)
+        {
+            if (string.IsNullOrWhiteSpace(hintName))
+            {
+                throw new ArgumentNullException(nameof(hintName));
+            }
+
+            if (!Path.GetExtension(hintName).Equals(".cs", StringComparison.OrdinalIgnoreCase))
+            {
+                hintName = string.Concat(hintName, ".cs");
+            }
+
+            if (AdditionalSources.Contains(hintName))
+            {
+                throw new ArgumentException(CodeAnalysisResources.HintNameUniquePerGenerator, nameof(hintName));
```

Comment:
Can we file a bug to track this? I don't think this is sustainable long term.

Reply:
Yes. Generators essentially represent a form of distributed development. Any aspect of the implementation which requires them to coordinate, such as not having conflicting hint names, makes authoring reliable generators difficult. The only safe approach would be to always use a GUID at which point it's not much of a hint.

Reply:
No, they only have to be unique *within* a generator. Different generators can have the same hint name.

---
In reply to: [434765041](https://github.com/dotnet/roslyn/pull/44768#discussion_r434765041) [](ancestors = 434765041)

Reply:
Gotcha. Assume we have tests for that?

Reply:
Yeah `Generator_HintName_MustBe_Unique` and `Generator_HintName_Is_Appended_With_GeneratorName`

---
In reply to: [434770864](https://github.com/dotnet/roslyn/pull/44768#discussion_r434770864) [](ancestors = 434770864)